web/book/src/internals/modules.md

```diff
@@ -0,0 +1,307 @@
+Such normalization is desired because it restrains the possible file system
+layouts to a comprehensible and predictable layout, while not sacrificing any
+functionality.
+
+Described importing rules don't achieve this "single normalized representation"
+in full, since any leaf modules could be replaced by a directory module with
+zero submodules, without any semantic changes. Restricting directory modules to
+have at least one sub-module would not improve approachability enough to justify
+adding this restriction.
+
+For example, the following module structure is annotated with files names in
+which the modules would reside:
+
+```prql no-eval
+module my_project           # _my_project.prql
+{
+    module sales {}         # sales.prql
+    module projections      # projections/_projections.prql
+    {
+        module year_2023 {} # projections/year_2023.prql
+        module year_2024 {} # projections/year_2024.prql
+    }
+}
+```
+
+If module `my_project.sales` wants to add a submodule `util`, it has to be
+converted to a directory modules. This means that it has to be moved to
+`sales/_sales.prql`. The new module would reside in `sales/util.prql`.
+
+The annotated layout is not the only possible layout for this module structure,
+since any of the modules `sales`, `year_2023` or `year_2024` could be converted
+into a directory module with zero sub-modules.
+
+Point 4 of design goals means that each declaration within a project has a
+single fully-qualified name within this project. This is insured by strict rules
+regarding importing files and the fact that the module structure is a tree.
+
+## Declaration order
+
+The order of declarations in a module holds no semantic value, except the "last
+`main` variable".
+
+References between modules can be cyclic.
+
+```
+module mod_a {
+    let decl_a_1 = ...
+    let decl_a_2 = (from mod_b.decl_b | take 10)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a | take 10)
+}
+```
+
+References between variable declarations cannot be cyclic.
+
+```
+let decl_a = (from decl_b)
+let decl_b = (from decl_a) # error: cyclic reference
+```
+
+```
+module mod_a {
+    let decl_a = (from mod_b.decl_b)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a) # error: cyclic reference
+}
+```
+
+## Compiler interface
+
+`prql-compiler` provides two interfaces for compiling files.
+
+**Multi-file interface** requires three arguments:
+
+- path to the file containing the module which is the root of the module
+  structure,
+- identifier of the pipeline that should be compiled to RQ and,
+- a "file loader", which can load files on-demand.
+
+The path to the root module can be automatically detected by searching for
+`.prql` files starting with `_` in the current working directory.
+
+Example prqlc usage:
+
+```
+$ prqlc compile _project.prql sales.projections.orders_2024
+$ prqlc compile sales.projections.orders_2024
+```
+
+**Single-file interface** requires a single argument; the PRQL source. Any
```

Comment:
Another way of conceptualizing this is that there's a new stage in the compiler before the current parser, where files are collected. It's possible to skip that stage by passing PRQL directly.

Reply:
Yes. That will fit better with the architecture of the compiler.